dbms/src/Common/BitpackingPrimitives.cpp

```diff
@@ -0,0 +1,317 @@
+    {
+        T value = unalignedLoad<T>(dst + i * sizeof(T));
+        value = static_cast<T>(value & ((TU(1) << width) - TU(1)));
+        T result = (value ^ mask) - mask;
+        Store(result, dst + i * sizeof(T));
+    }
+}
+
+template <typename T>
+UInt8 BitpackingPrimitives::getEffectiveWidth(UInt8 width)
+{
+    UInt8 bits_of_type = sizeof(T) * 8;
+    UInt8 type_size = sizeof(T);
+    if (width + type_size > bits_of_type)
+    {
+        return bits_of_type;
+    }
+    return width;
+}
+
+template <typename T>
+void BitpackingPrimitives::packGroup(unsigned char * dst, const T * values, UInt8 width)
+{
+    if constexpr (std::is_same<T, Int8>::value || std::is_same<T, UInt8>::value)
+    {
+        fastpforlib::fastpack(
+            reinterpret_cast<const UInt8 *>(values),
+            reinterpret_cast<UInt8 *>(dst),
+            static_cast<UInt32>(width));
+    }
+    else if constexpr (std::is_same<T, Int16>::value || std::is_same<T, UInt16>::value)
+    {
+        fastpforlib::fastpack(
+            reinterpret_cast<const UInt16 *>(values),
+            reinterpret_cast<UInt16 *>(dst),
+            static_cast<UInt32>(width));
+    }
+    else if constexpr (std::is_same<T, Int32>::value || std::is_same<T, UInt32>::value)
+    {
+        fastpforlib::fastpack(
+            reinterpret_cast<const UInt32 *>(values),
+            reinterpret_cast<UInt32 *>(dst),
+            static_cast<UInt32>(width));
+    }
+    else if constexpr (std::is_same<T, Int64>::value || std::is_same<T, UInt64>::value)
+    {
+        fastpforlib::fastpack(
+            reinterpret_cast<const UInt64 *>(values),
+            reinterpret_cast<UInt32 *>(dst),
+            static_cast<UInt32>(width));
+    }
+    else
+    {
+        throw Exception("Unsupported type for bitpacking");
+    }
+}
+
+template <typename T>
+void BitpackingPrimitives::unPackGroup(
+    unsigned char * dst,
+    const unsigned char * src,
+    UInt8 width,
+    bool skip_sign_extension)
+{
+    if constexpr (std::is_same<T, Int8>::value || std::is_same<T, UInt8>::value)
+    {
+        fastpforlib::fastunpack(
+            reinterpret_cast<const UInt8 *>(src),
+            reinterpret_cast<UInt8 *>(dst),
+            static_cast<UInt32>(width));
+    }
+    else if constexpr (std::is_same<T, Int16>::value || std::is_same<T, UInt16>::value)
+    {
+        fastpforlib::fastunpack(
+            reinterpret_cast<const UInt16 *>(src),
+            reinterpret_cast<UInt16 *>(dst),
+            static_cast<UInt32>(width));
+    }
+    else if constexpr (std::is_same<T, Int32>::value || std::is_same<T, UInt32>::value)
+    {
+        fastpforlib::fastunpack(
+            reinterpret_cast<const UInt32 *>(src),
+            reinterpret_cast<UInt32 *>(dst),
+            static_cast<UInt32>(width));
+    }
+    else if constexpr (std::is_same<T, Int64>::value || std::is_same<T, UInt64>::value)
+    {
+        fastpforlib::fastunpack(
+            reinterpret_cast<const UInt32 *>(src),
```

Comment:
Why UInt32 here?

Reply:
The lib will split 64 bits into two 32 bits